lib/srv/desktop/rdp/rdpclient/src/lib.rs

```diff
@@ -95,59 +96,84 @@ impl IronRDPClient {
 /// Client has an unusual lifecycle:
 /// - The function connect_rdp calls Client::new(), which creates it on the heap (Box::new), grabs a raw pointer(Box::into_raw),
 ///   and returns in to Go.
-/// - Most other exported rdp functions (pub unsafe extern "C") take the raw pointer and convert it (Box::leak(Box::from_raw(ptr)))
-///   to a reference (&'static mut Client), which can then be used without dropping the client.
+/// - Most other exported rdp functions (pub unsafe extern "C") take the raw pointer and convert it Client::from_raw
+///   to a reference (&Client), which can then be used without dropping the client.
 /// - The function free_rdp takes the raw pointer and drops it.
 ///
+/// The Client is forced to be Sync. See "Go/Rust Interface" in ../README.md for more details.
+///
 /// The Client makes use of asynchronous rust via the tokio runtime. A single runtime is created in connect_rdp and held on to by the
 /// Client. The exported rdp functions which need to call async rust functions start with `client.tokio_rt.handle().clone().block_on( ... )`,
 /// which creates a new task on the tokio runtime and blocks the current thread until it completes. Since these functions are called from
 /// Go, the "current thread" can be thought of as whichever goroutine the exported rdp function is called from. Because the client might
 /// be being used by multiple goroutines concurrently, it is up to the programmer to consider any synchronization mechanisms that might
 /// need to be implemented as features are added to the Client going forward.
 pub struct Client {
-    iron_rdp_client: IronRDPClient,
+    iron_rdp_client: Mutex<IronRDPClient>,
     tokio_rt: tokio::runtime::Runtime,
     go_ref: usize,
 }
 
+/// Forces the compiler to check that the Client struct is Sync.
+/// See the README for more details.
+const _: () = {
+    const fn assert_sync<T: Sync>() {}
+    assert_sync::<Client>();
+};
+
```

Comment:
I would actually add a `where Client: Sync` to the FFI functions that use the `*const Client` and `where Client: Send` to the drop function, so it's clear why the bound exists and why we need to assert it rather than letting Rust figure it out naturally - i.e. because those bounds are not enforced by the FFI conversion.

Reply:
Makes sense. I have a lot of refactoring piled up on top of this, so I'm going to add it as a separate PR at the end of that chain (rather than doing it here and then needing to resolve a bunch of conflicts).

Reply:
https://github.com/gravitational/teleport/pull/29830